src/api/providers/vscode-lm.ts

```diff
@@ -0,0 +1,313 @@
+     * @returns Promise resolving to the first matching language model chat instance
+     * @throws Error when no matching models are found with the given selector
+     * 
+     * @example
+     * const selector = { vendor: "copilot", family: "gpt-4o" };
+     * const chatClient = await createClient(selector);
+     */
+    async createClient(selector: vscode.LanguageModelChatSelector): Promise<vscode.LanguageModelChat> {
+
+        const models: vscode.LanguageModelChat[] = await vscode.lm.selectChatModels(selector);
+
+        if (models.length === 0) {
+
+            throw new Error("Cline <Language Model API>: No models were found whilst using the specified selector.");
+        }
+
+        return models[0];
+    }
+
+    /**
+     * Creates and streams a message using the VS Code Language Model API.
+     * 
+     * @param systemPrompt - The system prompt to initialize the conversation context
+     * @param messages - An array of message parameters following the Anthropic message format
+     * 
+     * @yields {ApiStream} An async generator that yields either text chunks or tool calls from the model response
+     * 
+     * @throws {Error} When vsCodeLmModelSelector option is not provided
+     * @throws {Error} When the response stream encounters an error
+     * 
+     * @remarks
+     * This method handles the initialization of the VS Code LM client if not already created,
+     * converts the messages to VS Code LM format, and streams the response chunks.
+     * Tool calls handling is currently a work in progress.
+     */
+    async *createMessage(systemPrompt: string, messages: Anthropic.Messages.MessageParam[]): ApiStream {
+
+        if (!this.client) {
+
+            if (!this.options.vsCodeLmModelSelector) {
+
+                throw new Error("Cline <Language Model API>: The 'vsCodeLmModelSelector' option is required for the 'vscode-lm' provider.");
+            }
+
+            this.client = await this.createClient(this.options.vsCodeLmModelSelector);
+        }
+
+        const vsCodeLmMessages: vscode.LanguageModelChatMessage[] = [
+            vscode.LanguageModelChatMessage.Assistant(systemPrompt),
+            ...convertToVsCodeLmMessages(messages),
+        ];
+
+        const response: vscode.LanguageModelChatResponse = await this.client.sendRequest(
+            vsCodeLmMessages,
+            { justification: `Cline would like to use '${this.client.name}', Click 'Allow' to proceed.` },
+            new vscode.CancellationTokenSource().token
+        );
+
+        try {
+
+            for await (const chunk of response.stream) {
+
+                if (chunk instanceof vscode.LanguageModelTextPart) {
+
+                    yield {
+                        type: "text",
+                        text: chunk.value,
+                    };
+                }
+
+                if (chunk instanceof vscode.LanguageModelToolCallPart) {
+
+                    // TODO: Determine how to handle tool calls.
```

Comment:
This todo does not mean that tool calling does not work.
When i tested this implementation cline could run my terminal commands just fine!

If i understand it correctly, the VSCode language model is able to call new tools. That's what this return type seems to be for. I think VSCode uses this for the "Copilot Edits" feature. But it seems to be reduntant within cline.

I'm thinking i might add some form of error toast, or nicely pass this called tool to cline or something.

Reply:
This code snippet has since been edited.